File: scripts/parse-eoir-pdf.py
#!/usr/bin/env python3
"""
Parse EOIR Accredited Representatives Roster PDF using pdfplumber
(table-aware extraction — preserves column structure).

Source: https://www.justice.gov/eoir/page/file/942311/dl
Output: output/eoir-immigration-reps.csv

~2,641 accredited immigration reps in the US (non-lawyer immigration
representatives who work at recognized nonprofit organizations).
"""

import csv
import os
import re
import sys
import pdfplumber

PDF_PATHS = [
    '/tmp/eoir-reps.pdf',
    '/tmp/eoir-1398081.bin',
]
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), '..', 'output')
OUTPUT_FILE = os.path.join(OUTPUT_DIR, 'eoir-immigration-reps.csv')


def parse_pdf(path):
    """Extract table rows from PDF using pdfplumber."""
    entries = []
    with pdfplumber.open(path) as pdf:
        print(f'  Pages: {len(pdf.pages)}')
        for page_num, page in enumerate(pdf.pages, 1):
            tables = page.extract_tables()
            if not tables:
                # Fallback: extract text and try to parse row-wise
                text = page.extract_text() or ''
                continue

            for table in tables:
                for row in table:
                    if row and any(cell for cell in row if cell):
                        entries.append(row)
            if page_num % 20 == 0:
                print(f'    Parsed {page_num}/{len(pdf.pages)} pages, {len(entries)} rows so far')
    return entries


def normalize_row(row):
    """Extract name, accreditation, org, dates from a table row."""
    # Clean cells
    cells = [((c or '').strip().replace('\n', ' ')) for c in row]

    # Skip header rows and empty rows
    if not any(cells): return None
    header_markers = {'name', 'accreditation', 'organization', 'address',
                      'expiration', 'expires', 'status', 'appointment'}
    if sum(1 for c in cells if c.lower() in header_markers) >= 2:
        return None

    # Heuristic: find name cell (typically first cell with Lastname, Firstname pattern)
    name_cell = ''
    for c in cells:
        if re.match(r'^[A-Z][a-zA-Z\-\' ]+,?\s+[A-Z][a-zA-Z\-\'\s]+$', c) and len(c) > 3:
            name_cell = c
            break
    if not name_cell and cells:
        name_cell = cells[0]

    # Extract accreditation, org, expiration from remaining cells
    accred = ''
    org = ''
    expires = ''
    address = ''

    for c in cells:
        if not c or c == name_cell: continue
        if 'partial' in c.lower() or 'full accreditation' in c.lower():
            accred = c
        elif re.match(r'^\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}', c):
            if not expires: expires = c
        elif len(c) > 10 and not org:
            org = c
        elif len(c) > 20 and org and not address:
            address = c

    # Parse first/last
    first_name, last_name = '', ''
    if ',' in name_cell:
        parts = name_cell.split(',', 1)
        last_name = parts[0].strip()
        first_name = parts[1].strip().replace('(DHS only)', '').strip()
    else:
        parts = name_cell.split()
        if len(parts) >= 2:
            first_name = parts[0]
            last_name = ' '.join(parts[1:])
        else:
            first_name = name_cell

    # DHS marker
    dhs_only = '(DHS only)' in name_cell or 'DHS' in name_cell
    first_name = first_name.replace('(DHS only)', '').strip()
    last_name = last_name.replace('(DHS only)', '').strip()

    if not first_name and not last_name:
        return None

    return {
        'first_name': first_name,
        'last_name': last_name,
        'accreditation': accred,
        'dhs_only': 'Y' if dhs_only else '',
        'organization': org,
        'address': address,
        'expires': expires,
        'raw': ' | '.join(cells)[:300],
    }


def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    print('=== EOIR Accredited Immigration Reps Parser (pdfplumber) ===')

    all_rows = []
    for path in PDF_PATHS:
        if not os.path.exists(path):
            print(f'  Skip: {path} (not found)')
            continue
        print(f'  Parsing {path}...')
        try:
            rows = parse_pdf(path)
            print(f'    Raw rows: {len(rows)}')
            all_rows.extend(rows)
        except Exception as e:
            print(f'  ERROR parsing {path}: {e}')

    # Normalize
    entries = []
    seen = set()
    for r in all_rows:
        e = normalize_row(r)
        if not e: continue
        key = (e['first_name'] + '|' + e['last_name'] + '|' + e['organization']).lower()
        if key in seen: continue
        seen.add(key)
        entries.append(e)

    # Write CSV
    cols = ['first_name', 'last_name', 'accreditation', 'dhs_only',
            'organization', 'address', 'expires', 'niche', 'source', 'country', 'raw']
    with open(OUTPUT_FILE, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=cols)
        writer.writeheader()
        for e in entries:
            e['niche'] = 'immigration'
            e['source'] = 'eoir_accredited_reps'
            e['country'] = 'US'
            writer.writerow(e)

    print(f'\n=== DONE ===')
    print(f'Total raw rows: {len(all_rows)}')
    print(f'Unique entries: {len(entries)}')
    print(f'With org:       {sum(1 for e in entries if e.get("organization"))}')
    print(f'Output: {OUTPUT_FILE}')

    # Sample
    print('\nSample (first 5):')
    for e in entries[:5]:
        print(f'  {e["first_name"]} {e["last_name"]} ({e.get("accreditation","?")}) — {e.get("organization","?")[:60]}')


if __name__ == '__main__':
    main()

File: scripts/parse-eoir-pdf.js
#!/usr/bin/env node
/**
 * Parse EOIR Accredited Representatives Roster PDF into CSV.
 *
 * Source: https://www.justice.gov/eoir/page/file/942311/dl
 * Output: output/eoir-immigration-reps.csv
 *
 * EOIR accredits non-lawyer immigration representatives who can represent
 * clients before immigration courts. These are immigration service providers
 * at recognized organizations (nonprofits, legal aid). ~2,641 individuals.
 *
 * Each entry in the PDF typically contains:
 *   - Name
 *   - Accreditation level (Partial / Full)
 *   - Organization name
 *   - Address
 *   - Expiration date
 */

const fs = require('fs');
const path = require('path');
const pdf = require('pdf-parse');

const PDF_PATHS = [
  '/tmp/eoir-reps.bin',
  '/tmp/eoir-1398081.bin',
];

const OUTPUT_DIR = path.join(__dirname, '..', 'output');
const OUTPUT_FILE = path.join(OUTPUT_DIR, 'eoir-immigration-reps.csv');

function escapeCsv(val) {
  if (val == null) return '';
  const s = String(val).trim();
  if (s.includes(',') || s.includes('"') || s.includes('\n')) {
    return '"' + s.replace(/"/g, '""') + '"';
  }
  return s;
}

async function parsePdf(filepath) {
  const buffer = fs.readFileSync(filepath);
  const data = await pdf(buffer);
  return data.text;
}

// Parse EOIR format. Typical entry block:
//   John Smith, Partial Accreditation
//   Catholic Charities of Atlanta
//   123 Main St
//   Atlanta, GA 30303
//   Expires: 10/31/2027
function parseEntries(text) {
  const lines = text.split(/\r?\n/).map(l => l.trim());
  const entries = [];

  // Try line-based parsing first — look for "Expires: MM/DD/YYYY" as entry terminator
  let currentBuffer = [];

  for (const line of lines) {
    if (!line) continue;

    // Skip header/footer noise
    if (/^(page \d+|accredited rep|number of|executive office|department of justice|u\.s\. department|updated)/i.test(line)) continue;
    if (/^\d+\s*$/.test(line)) continue;

    currentBuffer.push(line);

    // Entry boundary: line with "Expires" date OR paired "Partial"/"Full Accreditation"
    const isExpiry = /expires:?\s*\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}/i.test(line);
    const isAccred = /\b(Partial|Full)\s+Accreditation\b/i.test(line);

    if (isExpiry && currentBuffer.length >= 2) {
      entries.push(currentBuffer.join('\n'));
      currentBuffer = [];
    }
  }

  if (currentBuffer.length) entries.push(currentBuffer.join('\n'));

  // Now parse each entry
  const parsed = [];
  for (const raw of entries) {
    const lines = raw.split('\n').filter(Boolean);

    let firstName = '', lastName = '', accreditation = '';
    let organization = '', address = '', city = '', state = '', zip = '', expires = '';

    // Look for accreditation marker
    const accIdx = lines.findIndex(l => /\b(Partial|Full)\s+Accreditation\b/i.test(l));

    if (accIdx >= 0) {
      const accLine = lines[accIdx];
      const accMatch = accLine.match(/^(.+?)[,\s]+(Partial|Full)\s+Accreditation/i);
      if (accMatch) {
        const fullName = accMatch[1].trim();
        accreditation = accMatch[2];

        const parts = fullName.split(/\s+/).filter(Boolean);
        if (parts.length >= 2) {
          firstName = parts[0];
          lastName = parts.slice(1).join(' ');
        } else {
          firstName = fullName;
        }
      }

      // Org is usually the next non-empty line after accreditation
      if (accIdx + 1 < lines.length) organization = lines[accIdx + 1].trim();

      // Address: usually 2-3 lines after org
      const addrLines = [];
      for (let i = accIdx + 2; i < lines.length; i++) {
        if (/expires:/i.test(lines[i])) {
          const m = lines[i].match(/expires:?\s*(\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4})/i);
          if (m) expires = m[1];
          break;
        }
        addrLines.push(lines[i]);
      }
      address = addrLines.join(', ');

      // Try to extract city, state, zip from address
      const cszMatch = address.match(/([A-Z][a-zA-Z\s.\-]+),\s*([A-Z]{2})\s+(\d{5}(?:-\d{4})?)/);
      if (cszMatch) {
        city = cszMatch[1].trim();
        state = cszMatch[2];
        zip = cszMatch[3];
      }
    }

    if (!firstName && !organization) continue;

    parsed.push({
      first_name: firstName,
      last_name: lastName,
      accreditation,
      organization,
      address,
      city,
      state,
      zip,
      expires,
      raw_block: raw,
    });
  }

  return parsed;
}

async function main() {
  if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  console.log('=== EOIR Accredited Immigration Reps Parser ===');

  const allEntries = [];
  for (const p of PDF_PATHS) {
    if (!fs.existsSync(p)) {
      console.log(`  Skip (not found): ${p}`);
      continue;
    }
    console.log(`  Parsing ${p}...`);
    const text = await parsePdf(p);
    console.log(`    Text length: ${text.length} chars`);
    const entries = parseEntries(text);
    console.log(`    Parsed ${entries.length} entries`);
    allEntries.push(...entries);
  }

  // Dedup by name + org
  const seen = new Set();
  const unique = [];
  for (const e of allEntries) {
    const key = (e.first_name + '|' + e.last_name + '|' + e.organization).toLowerCase();
    if (!seen.has(key)) {
      seen.add(key);
      unique.push(e);
    }
  }

  // Write CSV
  const cols = [
    'first_name', 'last_name', 'accreditation', 'organization',
    'address', 'city', 'state', 'zip', 'expires',
    'niche', 'source', 'country',
  ];
  const header = cols.join(',');
  const rows = unique.map(e => cols.map(c => {
    if (c === 'niche') return escapeCsv('immigration');
    if (c === 'source') return escapeCsv('eoir_accredited_reps');
    if (c === 'country') return escapeCsv('US');
    return escapeCsv(e[c] || '');
  }).join(','));
  fs.writeFileSync(OUTPUT_FILE, header + '\n' + rows.join('\n') + '\n');

  console.log(`\n=== DONE ===`);
  console.log(`Total entries:  ${allEntries.length}`);
  console.log(`Unique:         ${unique.length}`);
  console.log(`With org:       ${unique.filter(e => e.organization).length}`);
  console.log(`With address:   ${unique.filter(e => e.address).length}`);
  console.log(`Output:         ${OUTPUT_FILE}`);

  // Sample
  console.log('\nSample (first 3):');
  for (const s of unique.slice(0, 3)) {
    console.log(`  ${s.first_name} ${s.last_name} (${s.accreditation}) — ${s.organization} — ${s.city}, ${s.state}`);
  }
}

main().catch(err => {
  console.error('FATAL:', err);
  process.exit(1);
});
